### ISSUE TITLE
Redirect to 'Redirect URLs' twice when enable 'Auto signin'

### ISSUE THREAD

**Issue by Sierting-Industrial** [casdoor/casdoor#1650]
# Environment
Python：3.10
Django: 4.1
casdoor-python-sdk: 1.8.0
# Problem
When enable the `Auto signin` and `Signin session`, if the user has signed-in, the page will redirect twice to callback URL with different `code` like picture below.
![](https://user-images.githubusercontent.com/64769975/224655352-f67cdbdc-3eb4-4751-8919-a0287a9b6084.jpg)
If disable the `Auto signin`, then there is no problem.
<img width="897" alt="image" src="https://user-images.githubusercontent.com/64769975/224660697-bda63dc5-2523-42e4-8632-35e001533e82.png">
It is preliminarily judged that the redirection code of front-end's JS caused the problem.

**Comment by Sierting-Industrial** [comment]
# More detail about problem
![img_v2_6d40a646-58a0-45f8-b333-9c56dad321ag](https://user-images.githubusercontent.com/64769975/224660897-3b378377-9d0d-4256-8d5c-71573fa5918d.jpg)

![img_v2_5a5e600b-f877-402e-bf55-99f41fc2738g](https://user-images.githubusercontent.com/64769975/224661012-4c8a36cc-7fee-4c27-8f7e-d2aab7ebc2ad.jpg)

This problem causes Django open two thread to handle this two callback, and when creating the new user, two thread pass through the `if` condition at the same time and create user, then because of one thread is already created the user, the other thread will raise the `IntegrityError` which means database already has the same user.

The right way is that Casdoor should only redirect once when `Auto signin` is enable. And only Chromium kernel browsers(Chrome, Edge) has this problem so far after I tested the Chrome,Safari and Edge.

**Comment by leo220yuyaodog** [comment]
@Sierting-Industrial The `localhost:8000/callback?***` is your application host? Casdoor servicer sends twice get request URL to you host?

**Comment by hsluoyz** [comment]
@Sierting-Industrial which frontend SDK do you use? I cannot reproduce your issue via Casnode: https://forum.casbin.com/

Casnode uses casdoor-js-sdk for frontend and casdoor-go-sdk for backend

![image](https://user-images.githubusercontent.com/3787410/225942495-8e9256d0-2953-49bf-9f70-c975adc642c9.png)

You can see there's only one signin request

![image](https://user-images.githubusercontent.com/3787410/225943188-bdd693f6-8055-4098-bffd-056d3faf65d5.png)

**Comment by Sierting-Industrial** [comment]
OK, I did not use the fronted sdk, I am using the basic way to login: user click login button, and application redirect user to Casdoor with some params like `http://10.0.8.114:8000/login/oauth/authorize?client_id=XXXXX&response_type=code&redirect_uri=http%3A%2F%2F127.0.0.1%3A8000%2Fcallback&scope=openid+profile+email+phone+username&state=tianshou`. I am developing locally so the callback URL is my localhost, this doesn't bother anything. 
About the python-sdk, I developed a Django plugin that could use with [Django-RESTFramework](https://www.django-rest-framework.org/), the main code as the below, or you can check the full code [here](https://github.com/Anion3r/djangorestframework-casdoor).
```python
# backends.py
from django.contrib.auth import get_user_model
from django.contrib.auth.backends import BaseBackend
from .settings import api_settings


class CasdoorBackend(BaseBackend):
    def authenticate(self, request, access_token=None, **kwargs):
        if access_token is None:
            access_token = kwargs.pop('access_token', None)
        if access_token is None:
            return None

        try:
            token = api_settings.sdk.parse_jwt_token(access_token)
        except Exception as e:
            return None
        user_model = get_user_model()
        try:
            query = {
                user_model.USERNAME_FIELD: token.get('name', '')
            }
            user = user_model.objects.get(**query)
        except user_model.DoesNotExist:
            if not api_settings.AUTO_CREATE_USER:
                return None
            query = {}
            fields = [field.name for field in user_model._meta.get_fields()]
            for k, v in token.items():
                if k in api_settings.CASDOOR_TO_AUTH_MODEL.keys() and api_settings.CASDOOR_TO_AUTH_MODEL[k] in fields:
                    query[api_settings.CASDOOR_TO_AUTH_MODEL[k]] = v
            user = user_model.objects.create_user(**query)
        return user

    def get_user(self, user_id):
        try:
            return get_user_model().objects.get(pk=user_id)
        except get_user_model().DoesNotExist:
            return None

```

I have recorded a video to show what happed（[Video Link](https://sierting.feishu.cn/file/YNtYbW0MCoNfROxqZ0XcZAg5nfg)）.

In the video, If I didn't login to Casdoor it will redirect perfectly to my Django, If I'm already logged in to Casdoor, then the problemn will accure. 

The reason is that two redirections both come in to Django, and because of the database has no information about this user, both theads will raise a DoesNotExist exception when executing the `user = user_model.objects.get(**query)`, so they both go into the exception part to create a new user. Then two threads will execute the code `user = user_model.objects.create_user(**query)`  at the same time. Although both threads execute this code at the same time, the database system itself has a thread lock when creating data, so the two threads are created one after the other when creating new users. Then, because the user name field in the user table is unique, one of the two threads will raise an IntegrityError exception.

And this all problem can be avoid by disabling the `Auto signin`.

**Comment by Sierting-Industrial** [comment]
Yes, you are right, it is nothing to do wtih Django. I suspected earlier that it was Django's problem, but after I captured the HTTP flow by Yakit, it seems like Casdoor's Javascript called some function twice, like picture below:
<img width="1791" alt="image" src="https://user-images.githubusercontent.com/64769975/226815287-657ebbc3-418d-452f-a6df-b974681754ca.png">
The blue rectangle is me that want to login via Casdoor, and it perfectly redirect to the login page. And the red one is what Casdoor requests the server side API. So this is the reason why I think it maybe the issue of Casdoor.

And I reproduced the issue with [https://demo.casdoor.com/](https://demo.casdoor.com/), looks like it realy is Casdoor's issue.
<img width="1781" alt="image" src="https://user-images.githubusercontent.com/64769975/226818453-cf4cca03-2209-44de-8170-45def4bff1b9.png">

<img width="1478" alt="image" src="https://user-images.githubusercontent.com/64769975/226819457-b147a4e3-1686-429e-9a42-485a04bff2ee.png">

To facilitate your reproduction and debugging, I have uploaded the source code of my local [Django project](https://sierting.feishu.cn/file/NgiybFY4FoyrH9x8ZHdcZMsgnmc) for testing for you

**Comment by Sierting-Industrial** [comment]
And I test on your official forum site [https://forum.casbin.com/](https://forum.casbin.com/), It has this issue, too.

<img width="1574" alt="image" src="https://user-images.githubusercontent.com/64769975/226826511-308b22f3-d361-4ea7-9f08-48f2fd6539fd.png">
<img width="1590" alt="image" src="https://user-images.githubusercontent.com/64769975/226826775-74d28c3f-9900-42c4-b953-60e8cfe5e0fa.png">

It called `/api/login` API twice and returned twice.

<img width="1557" alt="image" src="https://user-images.githubusercontent.com/64769975/226827026-9216e251-09f7-4e69-80ff-31df9f8f89b2.png">

To reduce this, you need to login [https://door.casdoor.com/](https://door.casdoor.com/) first, and then login [https://forum.casbin.com/](https://forum.casbin.com/), so that you will see two redirections.

**Comment by leo220yuyaodog** [comment]
@Sierting-Industrial plz review and test this PR #1682
